src/main/host/thread_ptrace.c

```diff
@@ -809,20 +831,67 @@ static void threadptrace_flushPtrs(Thread* base) {
 static void _threadptrace_doAttach(ThreadPtrace* thread) {
     debug("thread %i attaching to child %i", thread->base.tid, (int)thread->base.nativeTid);
     if (ptrace(PTRACE_ATTACH, thread->base.nativeTid, 0, 0) < 0) {
+        if (errno == ESRCH) {
+            // This *should* only happen if the thread group exited (due to
+            // signal or exit_group(2)).
+            debug("Task exited while unattached");
+            // Task died while we were unattached.
+            //
+            // If this task was the thread leader, _threadptrace_reapProcess
+            // should be called later, which will get the exit code.
+            // Otherwise, there's no way to get the exit code now.
+            thread->childState = THREAD_PTRACE_CHILD_STATE_EXITED;
+            return;
+        } else if (errno == EPERM) {
+            // Experimentally we can alternatively get EPERM here if the thread
+            // has disappeared. Haven't found any documented reason why.
+            debug("Got EPERM when attempting to reattach");
```

Comment:
I wonder if this could happen if the thread exited and then another process was started by the system with the same tid. Though I don't think that is likely if tids are monotonically increasing (until they wrap).

Reply:
Yeah, agreed on both counts - might explain it but seems unlikely.

I was thinking about this some more and think I'm going to try a different approach. In general when something happens that kills the entire process (namely `exit_group` being called or being killed by a signal), the threads die at unpredictable times. After iteratively adding more and more "exception handling" to handle the thread going away at different points, I think in principle it could happen in any ptrace or waitpid call. 

I'm thinking instead when we know the entire process is killed, we should not try resuming any other threads in the process again. e.g. I'll add an exit_group syscall handler that marks the process dead before making the syscall. When the Process code returns from the thread it can check if it's dead, and if so just do the bare minimum cleanup of other threads. We can also skip writing to the tid mutex in this case, avoiding potential trouble of writing to the `/proc/x/mem` file of a dead process.

Reply:
Ahh, there is mention of handing `exit_group` in #826, but we never got there (yet). :)